subprojects/language-java/src/main/java/org/gradle/api/tasks/JavaExec.java

```diff
@@ -528,6 +528,9 @@ public JavaExec copyTo(JavaForkOptions options) {
      */
     @Input
     public JavaVersion getJavaVersion() {
+        if (getJavaLauncher().isPresent()) {
+            return JavaVersion.toVersion(getJavaLauncher().get().getMetadata().getLanguageVersion());
+        }
```

Comment:
I don't think this makes sense. We already use `getEffectivrExecutable` below which does the same

Reply:
I did it because here we only use the `JavaVersion` parsing, while the other path requires to use the `JvmVersionDetector` and I found it made it clearer that toolchain usage will overwrite executable configuration.

Reply:
Right, `JvmVersionDetector` has caching in it, so no need to shortcircuit